### PR TITLE
feat: Add Hook to monitor network connectivity status

### LINKED ISSUE #6409: Create utility to subscribe to connectivity state in Javascript
Assuming that we will receive an offline/offline boolean status via some means (whether it be NetInfo or bridging the connectivity status from the host apps), this utility should allow us to:

- Get the current online/offline status within the Editor
- Provide updates to other components within the Editor when the online/offline status changes

### PR DESCRIPTION
**Pull request by dcalhoun**
## Related
- https://github.com/wordpress-mobile/gutenberg-mobile/pull/6417
- https://github.com/wordpress-mobile/WordPress-iOS/pull/22112
- https://github.com/wordpress-mobile/WordPress-Android/pull/19692

## What?

Add a React Hook for monitoring the network connection status via the bridge to
the host app.

## Why?

The network connection status will enable future features improving the UX when
a user's network connection changes, e.g. displaying status indicators.
Fixes https://github.com/wordpress-mobile/gutenberg-mobile/issues/6409. 

## How?

Add a `useIsConnected` React Hook that invokes bridge methods retrieving and
monitoring the network connection status from the host app. Relying upon the
host app will result in a single source of truth and mitigate conflicting
statuses.

## Testing Instructions

1. Checkout this branch within the `gutenberg-mobile` submodule. 
2. Start the Metro server via `npm run start:reset`. 
3. Apply the following patch: 
    <details><summary>Test Diff</summary>

    ```diff
    diff --git a/packages/block-editor/src/components/block-list/index.native.js b/packages/block-editor/src/components/block-list/index.native.js
    index 810e23e4c1..9ab21320cb 100644
    --- a/packages/block-editor/src/components/block-list/index.native.js
    +++ b/packages/block-editor/src/components/block-list/index.native.js
    @@ -1,7 +1,7 @@
    /**
     * External dependencies
     */
    -import { View, Platform, Pressable } from 'react-native';
    +import { View, Platform, Pressable, Text } from 'react-native';
    
    /**
     * WordPress dependencies
    @@ -15,6 +15,7 @@ import {
        alignmentHelpers,
    } from '@wordpress/components';
    import { __ } from '@wordpress/i18n';
    +import { useIsConnected } from '@wordpress/react-native-bridge';
    
    /**
     * Internal dependencies
    @@ -227,6 +228,7 @@ export default function BlockList( {
            headerToolbar.height +
            blockToolbar.height +
            ( isFloatingToolbarVisible ? floatingToolbar.height : 0 );
    +	const { isConnected } = useIsConnected();
    
        return (
            <View
    @@ -235,6 +237,20 @@ export default function BlockList( {
                onLayout={ onLayout }
                testID="block-list-wrapper"
            >
    +			<View
    +				style={ {
    +					backgroundColor: 'red',
    +				} }
    +			>
    +				<Text
    +					style={ {
    +						color: 'white',
    +						fontSize: 30,
    +					} }
    +				>
    +					Connected: { isConnected === null ? 'null' : isConnected }
    +				</Text>
    +			</View>
                { isRootList ? (
                    <BlockListProvider
                        value={ {

    ```

    </details>
4. Configure the app to load the local Metro server:
    - For iOS, install pods via `LOCAL_GUTENBERG=true bundle exec pod install`. 
    - For Android, set `localGutenbergMobilePath` within `local-builds.gradle`. 
7. Build and run the WordPress app. 
8. Launch the editor and verify the connection status displayed is accurate. 
9. Toggle airplane mode, verify the connection status displayed is accurate. 

### Testing Instructions for Keyboard

n/a, this internal tool introduces no user-facing changes.

## Screenshots or screencast 
n/a, this internal tool introduces no visible changes.

### CODE REVIEW

File: packages/react-native-editor/ios/GutenbergDemo/GutenbergViewController.swift
Comment (twstokes) [inline packages/react-native-editor/ios/GutenbergDemo/GutenbergViewController.swift]: Super-nit 😄 

```suggestion
        return true
```

File: packages/react-native-bridge/index.js
Comment (dcalhoun) [inline packages/react-native-bridge/index.js:193]: I'd like to move this `useIsConnected` and its related logic into a separate file to improve code comprehension and make testing easier. However, doing so would currently create a cyclic dependency. 

After merging this work, I hope to refactor the bridge to avoid cyclic dependencies and enable testing.

Comment (derekblank) [reply packages/react-native-bridge/index.js:193]: I feel we may want to consider adding more detail to this hook's name. In Jetpack, the concept of [isConnected](https://github.com/Automattic/jetpack/blob/55443e71e888ceaab9644a7d1527eaf7e44233d7/docs/rest-api.md?plain=1#L225) refers to a Jetpack site connection. Although it's a separate repo, being more descriptive in the hook name may help avoid confusion, and help also clarify "what" is connected. I had been using `isNetworkConnected` in a previous example, and even considered `isMobileNetworkConnected` to be extra clear. Open to other perspectives, however. If we do decide to update the name, it can happen in a follow-up PR.

**Review by derekblank — APPROVED** [review]
Looks great, and was working well on both host apps during my testing. I left a small note about possibly adding more detail to the name of the hook itself, but it's not a blocker. 🚀

File: packages/react-native-bridge/index.js
Comment (dcalhoun) [reply packages/react-native-bridge/index.js:193]: Good feedback. I agree we should update the name to more clearly communicate its intent: `isNetworkConnected`. I suggest we refactor this in a future PR alongside implementing tests.

**Comment by dcalhoun** [comment]
I reverted this work in https://github.com/WordPress/gutenberg/pull/56836 to avoid merging bridge changes into the WordPress host apps, as we do not plan to integrate a tagged release before the next WordPress host app release. Introducing bridge changes without a formal tagged release would've disrupted the release process. 

I'll open another PR to merge this work after the next WordPress host app release (23.9 on Dec 11).

**Pull request merged**
